neuralmonkey/encoders/attentive.py

```diff
@@ -0,0 +1,117 @@
+    each row of the matrix is computed using a different attention head. This
+    matrix is exposed as the ``temporal_states`` property (the time dimension
+    corresponds to the different attention heads). The ``output`` property
+    provides a flattened and, optionally, projected representation of this
+    matrix.
+    """
+
+    # pylint: disable=too-many-arguments
+    def __init__(self,
+                 name: str,
+                 input_sequence: Attendable,
+                 hidden_size: int,
+                 num_heads: int,
+                 output_size: int = None,
+                 state_proj_size: int = None,
+                 dropout_keep_prob: float = 1.0,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        """Initialize an instance of the encoder."""
+        check_argument_types()
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+        TemporalStatefulWithOutput.__init__(self)
+
+        self.input_sequence = input_sequence
+        self.hidden_size = hidden_size
+        self.num_heads = num_heads
+        self.output_size = output_size
+        self.state_proj_size = state_proj_size
+        self.dropout_keep_prob = dropout_keep_prob
+
+        if self.dropout_keep_prob <= 0.0 or self.dropout_keep_prob > 1.0:
+            raise ValueError("Dropout keep prob must be inside (0,1].")
+
+        with self.use_scope():
+            self.train_mode = tf.placeholder(tf.bool, [], "train_mode")
+    # pylint: enable=too-many-arguments
+
+    @tensor
+    def attention_states(self) -> tf.Tensor:
+        return dropout(get_attention_states(self.input_sequence),
+                       self.dropout_keep_prob,
+                       self.train_mode)
+
+    @tensor
+    def attention_weights(self) -> tf.Tensor:
+        mask = get_attention_mask(self.input_sequence)
+        hidden = tf.layers.dense(self.attention_states, units=self.hidden_size,
+                                 activation=tf.tanh, use_bias=False,
+                                 name="S1")
+        energies = tf.layers.dense(hidden, units=self.num_heads,
+                                   use_bias=False, name="S2")
+        # shape: [batch_size, max_time, num_heads]
+        weights = tf.nn.softmax(energies, dim=1)
+        if mask is not None:
+            weights *= tf.expand_dims(mask, -1)
+            weights /= tf.reduce_sum(weights, axis=1, keep_dims=True) + 1e-8
+
+        return weights
+
+    @tensor
+    def temporal_states(self) -> tf.Tensor:
+        states = self.attention_states
+        if self.state_proj_size is not None:
+            states = tf.layers.dense(states, units=self.state_proj_size,
+                                     name="state_projection")
+
+        return tf.matmul(a=self.attention_weights, b=states, transpose_a=True)
+
+    @tensor
+    def temporal_mask(self) -> tf.Tensor:
+        return tf.ones(tf.shape(self.temporal_states)[:2], tf.float32)
```

Comment:
nechceš tady spíš zkopírovat masku tý input sequence? tohle udělá všechny pozice platný.

Reply:
Ne – jak je napsáno v docstringu, ty `temporal_states` mají fixní velikost a "časový rozměr" neodpovídá té vstupní sekvenci, ale jednotlivým attention heads. To je kvůli tomu, aby se pak za to dal připojit dekodér, který přes tenhle rozměr dělá další attention.

Reply:
jo aha, už to čtu.. ok